source/sigv4.c

```diff
@@ -794,6 +899,332 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+        assert( headerCount > 0 );
+
+        for( noOfHeaders = 0; noOfHeaders < headerCount; noOfHeaders++ )
+        {
+            assert( canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData != NULL );
+            sigV4Status = writeCanonicalHeaderToCanonicalRequest( noOfHeaders, canonicalRequest );
+
+            if( sigV4Status != SigV4Success )
+            {
+                break;
+            }
+        }
+
+        return sigV4Status;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static SigV4Status_t appendAllHeadersToCanonicalRequest( const char * pHeaders,
+                                                             size_t headersLen,
+                                                             uint32_t flags,
+                                                             CanonicalContext_t * canonicalRequest )
+    {
+        const char * start;
+        const char * end;
+        size_t keyFlag = 1, noOfHeaders = 0, i = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( pHeaders != NULL );
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        start = end = pHeaders;
+
+        for( i = 0; i < headersLen; i++ )
+        {
+            if( noOfHeaders == SIGV4_MAX_HTTP_HEADER_COUNT )
+            {
+                sigV4Status = SigV4MaxHeaderPairCountExceeded;
+                break;
+            }
+            /* Extracting each header key and value from the headers string. */
+            else if( ( keyFlag == 1 ) && ( pHeaders[ i ] == ':' ) )
+            {
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData = start;
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].key.dataLen = ( end - start );
+                start = end + 1U;
+                keyFlag = 0;
+            }
+            else if( ( keyFlag == 0 ) && ( !( flags & SIGV4_HTTP_HEADERS_ARE_CANONICAL_FLAG ) && ( pHeaders[ i ] == '\r' ) && ( ( i + 1 ) < headersLen ) && ( pHeaders[ i + 1 ] == '\n' ) ) )
+            {
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].value.pData = start;
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].value.dataLen = ( end - start );
+                start = end + 2U;
+                keyFlag = 1;
+                noOfHeaders++;
+            }
+            else if( ( keyFlag == 0 ) && ( ( flags & SIGV4_HTTP_HEADERS_ARE_CANONICAL_FLAG ) && ( pHeaders[ i ] == '\n' ) ) )
+            {
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].value.pData = start;
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].value.dataLen = ( end - start );
+                start = end + 1U;
+                keyFlag = 1;
+                noOfHeaders++;
+            }
+
+            end++;
+        }
```

Comment:
To reduce the complexity of this function, the logic of parsing the header key-value pairs can be culled out into a separate function. Here's my suggestion of the logic with some variable naming improvements.

```
SigV4Status_t parseHeaderKeyValueEntries( const char * pHeadersData,
                                          size_t headersDataLen,
                                          uint32_t flags,
                                          CanonicalContext_t * pCanonicalRequest )
{
    assert( pHeadersData != NULL );
    assertheadersDataLen > 0 );
    assert( pCanonicalRequest != NULL );

    size_t index = 0;
    const char * pKeyOrValStartLoc = pHeadersData;
    const char * pCurrLoc = pHeadersData;
    bool keyFound = false;

    for( index = 0; index < assertheadersDataLen; i++ )
    {
        pCurrLoc = &pHeadersData[index];
        if( noOfHeaders == SIGV4_MAX_HTTP_HEADER_COUNT )
        {
            sigV4Status = SigV4MaxHeaderPairCountExceeded;
            break;
        }
        /* Look for key part of an header field entry. */
        else if( ( keyFound == false ) && ( pHeaders[ index ] == ':' ) )
        {
            pCanonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData = pKeyOrValStartLoc;
            pCanonicalRequest->pHeadersLoc[ noOfHeaders ].key.dataLen = ( pCurrLoc - pKeyOrValStartLoc );
            pKeyOrValStartLoc = pCurrLoc + 1U;
            keyFound = false;
        }
        /* Look for header value part of a header field entry for both canonicalized and non-canonicalized forms. */
        else if( keyFound == true)
        {
            /* Non-canonicalized headers will have header values ending with "\r\n". */
            if( !( flags & SIGV4_HTTP_HEADERS_ARE_CANONICAL_FLAG ) && ( ( index + 1 ) < headersLen ) 
                    && (0 == strncmp(pCurrLoc, "\r\n", strlen("\r\n"))) ) 
            {
                pCanonicalRequest->pHeadersLoc[ noOfHeaders ].value.pData = pKeyOrValStartLoc;
                pCanonicalRequest->pHeadersLoc[ noOfHeaders ].value.dataLen = ( pCurrLoc - pKeyOrValStartLoc );
                /* Set starting location of the next header key string after the "\r\n". */
                pKeyOrValStartLoc = pCurrLoc + 2U;
                keyFound = false;
                noOfHeaders++;
            }
            /* Canonicalized headers will have header values ending just with "\n". */
            else if( !( flags & SIGV4_HTTP_HEADERS_ARE_CANONICAL_FLAG ) && ( pHeaders[ index ] == '\n' ) )
            {
                pCanonicalRequest->pHeadersLoc[ noOfHeaders ].value.pData = pKeyOrValStartLoc;
                pCanonicalRequest->pHeadersLoc[ noOfHeaders ].value.dataLen = ( pCurrLoc - pKeyOrValStartLoc );
                /* Set starting location of the next header key string after the "\n". */
                pKeyOrValStartLoc = pCurrLoc + 1U;
                keyFound = false;
                noOfHeaders++;
            }
            else
            {
                sigV4Status = Sigv4BadParameter;
                break;
            }
        }
    }

    return sigV4Status;
}
```

Reply:
`pCurrLoc = &pHeadersData[index];` . This line will update the currentLoc pointer every time but we want to update it only when either we reach the end of key or value.

Reply:
```
else
            {
                sigV4Status = Sigv4BadParameter;
                break;
            }
```

This part necessarily need not be always Invalid Parameter. It can be any other character in the value too.